bsg_mem/bsg_mem_1rw_sync_subbanked.v

```
@@ -0,0 +1,125 @@
+/*
+ *  bsg_mem_1rw_sync_subbanked.v
+ *
+ *  This module has the same interface/functionality as
+    bsg_mem_1rw_sync.
+ *
+ *  - width_p : width of the total memory
+ *  - els_p   : depth of the total memory
+ *
+ *  - num_subbank_p : Number of banks for the memory's width. width_p has
+ *                    to be a multiple of this number.
+ *
+ */
+
+  // For subbanked SRAMs, all subbanks must either read or write from the same address at once. 
+  // There could be an implementation to support independent subbank r/w by using a 1r1w backing SRAM. 
+  // This may make sense in an FPGA environment, but we leave this to future work.
+
+`include "bsg_defines.v"
+
+module bsg_mem_1rw_sync_subbanked
+  #(parameter `BSG_INV_PARAM(width_p)
+    , parameter `BSG_INV_PARAM(els_p)
+    , parameter `BSG_INV_PARAM(latch_last_read_p)
+    , parameter `BSG_INV_PARAM(num_subbank_p)
+
+      // Don't support depth subbanks due to conflicts
+    , localparam subbank_width_lp = width_p/num_subbank_p
+    , localparam mask_width_lp     = subbank_width_lp >>3
+    , localparam els_lp = `BSG_SAFE_CLOG2(els_p)
+  )
+  (   input clk_i
+  	, input reset_i
+   	, input [num_subbank_p-1:0] v_i
+  	, input w_i
+    , input [num_subbank_p-1:0][mask_width_lp-1:0] w_mask_i
+  	, input [num_subbank_p-1:0][subbank_width_lp-1:0] data_i
+  	, input [els_lp-1:0] addr_i
+  	, output logic [num_subbank_p-1:0][subbank_width_lp-1:0] data_o
+  );
+
+    logic [num_subbank_p-1:0][subbank_width_lp-1:0] data_lo;
+
+    if (num_subbank_p == 1) begin: no_mask_sram
+
+      bsg_mem_1rw_sync #(
+        .width_p(width_p)
+        ,.els_p(els_p)
+        ,.latch_last_read_p(0)
+      ) 
+      bank 
+      ( .clk_i(clk_i)
+        ,.reset_i(reset_i)
+        ,.v_i(|v_i)
+        ,.w_i(|w_i)
+        ,.addr_i(addr_i)
+        ,.data_i(data_i)
+        ,.data_o(data_lo)
+      );
+
+    end //no_mask_sram
+
+    if (num_subbank_p > 1) begin: byte_mask_sram
```

instead of "if (num_subbank_p > 1)", "else" should be enough?

---

Yeah, it would work but then we have to implement a check for num_subbank_p!=0. Otherwise our code will break.

---

Well, right now, if num_subbank_p is set to 0, wouldn't it just silently fail?

---

Yes, it would be failing on 0.